src/awkward/_connect/numba/layoutbuilder.py

```diff
@@ -0,0 +1,2547 @@
+            val._contents,
+            val._parameters,
+        )
+
+    elif isinstance(val, Unmasked):
+        return UnmaskedType(val._content, val._parameters)
+
+    else:
+        raise TypeError("unrecognized LayoutBuilder type")
+
+
+@numba.extending.overload(len)
+def LayoutBuilderType_len(builder):
+    if isinstance(
+        builder,
+        (
+            BitMaskedType,
+            ByteMaskedType,
+            EmptyType,
+            IndexedOptionType,
+            ListOffsetType,
+            NumpyType,
+            RecordType,
+            RegularType,
+            TupleType,
+            UnionType,
+            UnmaskedType,
+        ),
+    ):
+
+        def len_impl(builder):
+            return builder._length_get()
+
+        return len_impl
+
+
+########## Numpy ############################################################
+
+
+@final
+class Numpy(LayoutBuilder):
+    def __init__(self, dtype, *, parameters=None, initial=1024, resize=8.0):
+        self._data = GrowableBuffer(dtype=dtype, initial=initial, resize=resize)
+        self._parameters = parameters
+
+    @classmethod
+    def _from_buffer(cls, data):
+        out = cls.__new__(cls)
+        out._data = data
+        out._parameters = None
+        return out
+
+    def __repr__(self):
+        return f"ak.numba.lb.Numpy({self._data.dtype}, parameters={self._parameters})"
+
+    def numbatype(self):
+        return NumpyType(
+            numba.from_dtype(self.dtype), numba.types.StringLiteral(self._parameters)
+        )
+
+    @property
+    def _length(self):
+        return len(self._data)
+
+    def __len__(self):
+        return self._length
+
+    @property
+    def dtype(self):
+        return self._data.dtype
+
+    @property
+    def parameters(self):
+        return self._parameters
+
+    def append(self, x):
+        self._data.append(x)
+
+    def extend(self, data):
+        self._data.extend(data)
+
+    def clear(self):
+        self._data.clear()
+
+    def is_valid(self, error: str):
+        return True
+
+    def snapshot(self) -> ArrayLike:
+        """
+        Converts the currently accumulated data into an #ak.Array.
+        """
+        return ak.Array(
+            ak.contents.NumpyArray(self._data.snapshot(), parameters=self._parameters)
+        )
+
+
+class NumpyType(numba.types.Type):
+    def __init__(self, dtype, parameters):
+        super().__init__(
+            name=f"ak.numba.lb.Numpy({dtype}, parameters={parameters.literal_value if isinstance(parameters, numba.types.Literal) else None})"
+        )
+        self._dtype = dtype
+        self._parameters = parameters
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    @property
+    def parameters(self):
+        return numba.types.StringLiteral(self._parameters)
+
+    @property
+    def data(self):
+        return ak.numba.GrowableBufferType(self._dtype)
+
+    @property
+    def length(self):
+        return numba.types.float64
```

Comment:
These Numba type classes would be easier to define if you inherit your implementations in `LayoutBuilderType`.

```suggestion
class NumpyType(LayoutBuilderType):
    def __init__(self, dtype, parameters):
        super().__init__(
            name=f"ak.lb.Numpy({dtype!r}, parameters={parameters!r})"
        )
        self._dtype = dtype
        self._init(parameters)

    @property
    def dtype(self):
        return self._dtype

    @property
    def data(self):
        return ak.numba.GrowableBufferType(self._dtype)
```

(And similarly with all other Numba type classes for `LayoutBuilder` subclasses.)

Reply:
Oh, I made the type disambiguator `"ak.lb.Numpy"` because this doesn't need to be the literal path in Awkward (it wasn't, anyway), it just needs to be distinct enough that no other Numba extensions are likely to use it. We've claimed the `"ak."` prefix, and there's no need to follow that up with "`numba.`" since this string is only used in Numba. `"lb."` will be enough to clarify that these are the LayoutBuilder part of Awkward's Numba extensions.